mesonbuild/ast/interpreter.py

```diff
@@ -215,45 +230,91 @@ def evaluate_foreach(self, node):
     def evaluate_if(self, node):
         for i in node.ifs:
             self.evaluate_codeblock(i.block)
-        if not isinstance(node.elseblock, mparser.EmptyNode):
+        if not isinstance(node.elseblock, EmptyNode):
             self.evaluate_codeblock(node.elseblock)
 
     def get_variable(self, varname):
         return 0
 
     def assignment(self, node):
-        assert(isinstance(node, mparser.AssignmentNode))
+        assert(isinstance(node, AssignmentNode))
         self.assignments[node.var_name] = [node.value] # Save a reference to the value node
         if hasattr(node.value, 'ast_id'):
             self.reverse_assignment[node.value.ast_id] = node
         self.assign_vals[node.var_name] = [self.evaluate_statement(node.value)] # Evaluate the value just in case
 
-    def flatten_args(self, args, include_unknown_args: bool = False):
-        # Resolve mparser.ArrayNode if needed
+    def flatten_args(self, args: Any, include_unknown_args: bool = False, id_loop_detect: List[str] = []) -> List[str]:
```

Comment:
> id_loop_detect: List[str] = []

This is bad, don't make default arguments mutable. The way python implements default arguments is that the function object created by def stores a reference and passes that reference in if there is not argument. This means that if id_loop_detect is ever mutated, all subsequent calls will receive that mutated instance. This is a way to get really subtle and hard to find bugs. In fact, I see a source of such bugs in this function.

it's a recipe for disaster, you should use the Optional pattern here:
```python
def foo(id_loop: Optiona[List[str]] = None):
  if id_loop is None:
    id_loop = []
```

Reply:
Wow... Why would anyone think that implementing default arguments this way is a good idea? Sometimes I really wish meson was implemented in C++ then I would at least know exactly what is happening.

Anyway, fixed with the last commit. However, there still other locations in the meson codebase that use list default values (found by searching for `=[ ]*\[[ ]*\].*\):` in VS Code)

Reply:
I'll fix those. It's one of those hard cases. You basically either get to pass a callable for default arguments and eat the cost calling said callable or you get this sort of problem.